helix-core/src/test/java/org/apache/helix/manager/zk/TestZkHelixAdmin.java

```diff
@@ -1008,4 +1008,58 @@ public void testUpdateCustomizedStateConfig() throws Exception {
     Assert.assertEquals(listTypesFromZk.get(0), "mockType2");
     Assert.assertEquals(listTypesFromZk.get(1), "mockType3");
   }
+
+  @Test
+  public void testPurgeOfflineInstances() {
+    String className = TestHelper.getTestClassName();
+    String methodName = TestHelper.getTestMethodName();
+    String clusterName = className + "_" + methodName;
+
+    System.out.println("START " + clusterName + " at " + new Date(System.currentTimeMillis()));
+
+    HelixAdmin tool = new ZKHelixAdmin(_gZkClient);
+    tool.addCluster(clusterName, true);
+
+    HelixDataAccessor dataAccessor = new ZKHelixDataAccessor(clusterName, _baseAccessor);
+    PropertyKey.Builder keyBuilder = dataAccessor.keyBuilder();
+
+    // set default offline node timeout for purge in cluster config
+    ClusterConfig clusterConfig = dataAccessor.getProperty(keyBuilder.clusterConfig());
+    clusterConfig.setOfflineNodeTimeOutForPurge(100000L);
+    dataAccessor.setProperty(keyBuilder.clusterConfig(), clusterConfig);
+
+    String hostname = "host1";
+    String port = "9999";
+    String instanceName = hostname + "_" + port;
+    InstanceConfig config = new InstanceConfig(instanceName);
+    config.setHostName(hostname);
+    config.setPort(port);
+    tool.addInstance(clusterName, config);
+    tool.enableInstance(clusterName, instanceName, true);
+
+    LiveInstance liveInstance = new LiveInstance(instanceName);
+    liveInstance.setSessionId(UUID.randomUUID().toString());
+    liveInstance.setHelixVersion(UUID.randomUUID().toString());
+    dataAccessor.setProperty(keyBuilder.liveInstance(instanceName), liveInstance);
+
+    dataAccessor.removeProperty(keyBuilder.liveInstance(instanceName));
+    ZNRecord znRecord = new ZNRecord(instanceName);
+    znRecord
+        .setSimpleField("LAST_OFFLINE_TIME", String.valueOf(System.currentTimeMillis() - 50000L));
+    _baseAccessor.set(PropertyPathBuilder.instanceHistory(clusterName, instanceName), znRecord, 1);
+
+    // This purge will not remove the instance since the default timeout is not met yet.
+    tool.purgeOfflineInstances(clusterName, -1);
+    Assert.assertTrue(_gZkClient.exists(keyBuilder.instanceConfig(instanceName).getPath()),
+        "Instance should still be there");
+
+    // This purge will remove the instance as the customized purge timeout is met.
+    Map<String, String> purgeMap = new HashMap<>();
+    tool.purgeOfflineInstances(clusterName, 10000L);
+    Assert.assertFalse(_gZkClient.exists(keyBuilder.instanceConfig(instanceName).getPath()),
+        "Instance should already be dropped");
+
+    tool.dropCluster(clusterName);
+    System.out.println("END " + clusterName + " at " + new Date(System.currentTimeMillis()));
```

Comment:
important: should we clean up all zk resources after the test is done? e.g.) `tool.close()`

Reply:
All these close parts are already handled in the aftersuite of base test class.